packages/apollo-server-plugin-full-query-cache/src/ApolloServerPluginFullQueryCache.ts

```diff
@@ -0,0 +1,273 @@
+  //   key tagged with the mode "authenticated public".
+  //
+  // This allows the cache to provide different "public" results to anonymous
+  // users and logged in users ("no session" vs "authenticated public").
+  //
+  // A common implementation of this hook would be to look in
+  // requestContext.request.http.headers for a specific authentication header or
+  // cookie.
+  //
+  // This hook may return a promise because, for example, you might need to
+  // validate a cookie against an external service.
+  sessionId?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): ValueOrPromise<string | null>;
+
+  // Define this hook if you want the cache key to vary based on some aspect of
+  // the request other than the query document, operation name, variables, and
+  // session ID. For example, responses that include translatable text may want
+  // to return a string derived from
+  // requestContext.request.http.headers.get('Accept-Language'). The data may
+  // be anything that can be JSON-stringified.
+  extraCacheKeyData?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): ValueOrPromise<any>;
+
+  // If this hook is defined and returns false, the plugin will not read
+  // responses from the cache.
+  readFromCache?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): ValueOrPromise<boolean>;
+
+  // If this hook is defined and returns false, the plugin will not write the
+  // response to the cache.
+  writeToCache?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): ValueOrPromise<boolean>;
+}
+
+enum SessionMode {
+  NoSession,
+  Private,
+  AuthenticatedPublic,
+}
+
+function sha(s: string) {
+  return createHash('sha256')
+    .update(s)
+    .digest('hex');
+}
+
+interface BaseCacheKey {
+  document: string;
+  operationName: string | null;
+  variables: { [name: string]: any };
+  extra: any;
+}
+
+interface ContextualCacheKey {
+  sessionMode: SessionMode;
+  sessionId?: string | null;
+}
+
+type CacheKey = BaseCacheKey & ContextualCacheKey;
+
+function cacheKeyString(key: CacheKey) {
+  return sha(JSON.stringify(key));
+}
+
+export default function plugin(
+  options: Options = Object.create(null),
+): ApolloServerPlugin {
+  return {
+    requestDidStart(
+      outerRequestContext: GraphQLRequestContext<any>,
+    ): GraphQLRequestListener<any> {
+      const cache = new PrefixingKeyValueCache(
+        options.cache || outerRequestContext.cache!,
+        'fqc:',
+      );
+
+      let sessionId: string | null = null;
+      let baseCacheKey: BaseCacheKey | null = null;
+
+      async function cacheGet(
+        contextualCacheKeyFields: ContextualCacheKey,
+      ): Promise<GraphQLResponse | null> {
+        const key = cacheKeyString({
+          ...baseCacheKey!,
+          ...contextualCacheKeyFields,
+        });
+        const value = await cache.get(key);
+        if (value === undefined) {
+          return null;
+        }
+        return JSON.parse(value);
```

Comment:
Note that this currently will not return a valid `GraphQLResponse`, because the `errors` property is expected to contain instances of `GraphQLError`, not formatted errors. I've been looking at changing this to `GraphQLFormattedError` however, because I'm running into a similar problem in the gateway where we want to propagate errors from underlying services. So this gives me another data point in favor of making that change.

Reply:
I'll have a look at that today, to get a better sense of how much work it would be. Ideally we'd get it done this week as well, because I also need it for the gateway work. My main concern is that it's potentially breaking for people who have already adopted the plugin API. So I want to hear what Jesse thinks about that as well.

Reply:
So just to be clear, the issue is that GraphQLResponse isn't actually a json roundtrippable object?

Reply:
Yep

Reply:
It's also more consistent with how the `errors` field of a response is described in the spec: 
https://facebook.github.io/graphql/draft/#sec-Errors. So the idea is that we'd add a separate `didEncounterErrors` hook for error reporting, that will receive `GraphQLError` instances. But the response would contain the errors after applying `formatError`.